src/main/kotlin/io/github/inductiveautomation/kindling/utils/Diff.kt

```diff
@@ -0,0 +1,209 @@
+        preChange: List<T>,
+        postChange: List<T>,
+        equalityPredicate: (T, T) -> Boolean,
+        lcs: List<T> = lcs(preChange, postChange, equalityPredicate),
+    ) : this(
+        preChange,
+        postChange,
+        additions = postChange.mapIndexedNotNull { index, item ->
+            if (item in lcs) null else Diff(DiffType.ADDITION, item, index)
+        }.sortedBy { it.index },
+        deletions = preChange.mapIndexedNotNull { index, item ->
+            if (item in lcs) null else Diff(DiffType.DELETION, item, index)
+        }.sortedBy { it.index },
+    )
+
+    constructor(
+        pre: List<T>,
+        post: List<T>,
+        equalityPredicate: (T, T) -> Boolean,
+    ) : this(pre, post, equalityPredicate, lcs(pre, post, equalityPredicate))
+
+    val leftDiffList = buildList {
+        addAll(
+            preChange.mapIndexed { index, s ->
+                deletions.find { it.index == index } ?: Diff(DiffType.NOCHANGE, s, index)
+            },
+        )
+
+        var offset = 0
+        additions.forEach {
+            val existingDeletion: Diff<T> = get(it.index)
+            if (existingDeletion.type != DiffType.DELETION) {
+                add(it.index, it)
+                offset++
+            }
+        }
+    }
+
+    val rightDiffList = buildList {
+        addAll(
+            postChange.mapIndexed { index, s: T ->
+                additions.find { it.index == index } ?: Diff(DiffType.NOCHANGE, s, index)
+            },
+        )
+
+        deletions.forEach {
+            val diffInList: Diff<T> = get(it.index)
+
+            if (diffInList.type != DiffType.ADDITION) {
+                add(it.index, it)
+            }
+        }
+    }
+
+    val unifiedDiffList: DiffList<T> = buildList {
+        val leftList = leftDiffList.iterator()
+        val rightList = rightDiffList.iterator()
+
+        var leftItem = leftList.nextOrNull()
+        var rightItem = rightList.nextOrNull()
+
+        while (leftItem != null || rightItem != null) {
+            while (leftItem is Diff.Addition) {
+                leftItem = leftList.nextOrNull()
+            }
+
+            while (rightItem is Diff.Deletion) {
+                rightItem = rightList.nextOrNull()
+            }
+
+            when {
+                leftItem is Diff.NoChange && rightItem is Diff.NoChange -> {
+                    add(Diff(leftItem.type, leftItem.value, leftItem.index, rightItem.index))
+                    leftItem = leftList.nextOrNull()
+                    rightItem = rightList.nextOrNull()
+                }
+
+                leftItem is Diff.Deletion -> {
+                    add(leftItem)
+                    leftItem = leftList.nextOrNull()
+                }
+
+                rightItem is Diff.Addition -> {
+                    add(rightItem)
+                    rightItem = rightList.nextOrNull()
+                }
+            }
+        }
+    }
+}
+
+enum class DiffType(val symbol: String) {
```

Comment:
Since you already have a sealed interface above, you shouldn't need an enum as well. Just put the symbol (or, better, the formatting operations) as abstract methods on your sealed interface. Then you remove the need for polymorphic dispatch in DiffView.